### PR TITLE
[Macros] Expression macro as caller-side default argument

### PR DESCRIPTION
**Pull request by ApolloZhu**
Implementation for [[Pitch] Expression macro as caller-side default argument](https://forums.swift.org/t/pitch-expression-macro-as-caller-side-default-argument/69019)

Corresponding Swift Syntax PR: https://github.com/apple/swift-syntax/pull/2447

Resolves rdar://112031348

### CODE REVIEW

**Comment by ApolloZhu** [comment]
@swift-ci please smoke test

**Comment by ApolloZhu** [comment]
@swift-ci please smoke test

**Comment by ApolloZhu** [comment]
@swift-ci please smoke test

**Comment by ApolloZhu** [comment]
@swift-ci please smoke test

**Comment by ApolloZhu** [comment]
https://github.com/apple/swift-syntax/pull/2447
@swift-ci Please smoke test

File: lib/Sema/TypeCheckStmt.cpp [hunk 1]
```diff
@@ -119,7 +119,12 @@ namespace {
 
       // Caller-side default arguments need their @autoclosures checked.
       if (auto *DAE = dyn_cast<DefaultArgumentExpr>(E))
-        if (DAE->isCallerSide() && DAE->getParamDecl()->isAutoClosure())
+        if (DAE->isCallerSide() &&
+            (DAE->getParamDecl()->isAutoClosure() ||
+             (DAE->getParamDecl()->getDefaultArgumentKind() ==
+                  DefaultArgumentKind::ExpressionMacro &&
+              ParentDC->getASTContext().LangOpts.hasFeature(
```
Comment (DougGregor) [inline lib/Sema/TypeCheckStmt.cpp:126]: I don't think we need this `hasFeature` check here. Checking at the definition site is good enough.

**Review by DougGregor — COMMENTED** [review]
This is looking very good. There are a few tests we still need:

* One that serializes a public function using a default argument that's a macro expansion, then calls it from another module
* One that prints an API with a public function containing such a default argument into a .swiftinterface fiel
* One that checks the access control restrictions on the macros that can be used from a default argument

**Comment by DougGregor** [comment]
https://github.com/apple/swift-syntax/pull/2447
@swift-ci Please build toolchain

**Comment by DougGregor** [comment]
This will also make `#isolation` usable as a default argument without further work

**Comment by DougGregor** [comment]
https://github.com/apple/swift-syntax/pull/2447
@swift-ci Please smoke test

**Comment by ApolloZhu** [comment]
https://github.com/apple/swift-syntax/pull/2447
@swift-ci Please smoke test

**Comment by ApolloZhu** [comment]
https://github.com/apple/swift-syntax/pull/2447
@swift-ci Please build toolchain

File: lib/AST/ASTScopeCreation.cpp [hunk 2]
```diff
@@ -274,6 +274,14 @@ ASTSourceFileScope::ASTSourceFileScope(SourceFile *SF,
 
     // Determine the parent source location based on the macro role.
     AbstractFunctionDecl *bodyForDecl = nullptr;
+
+    if (SF->Kind == SourceFileKind::DefaultArgument) {
+      auto genInfo = *SF->getASTContext().SourceMgr.getGeneratedSourceInfo(
+          *SF->getBufferID());
+      parentLoc = ASTNode::getFromOpaqueValue(genInfo.astNode).getStartLoc();
+      goto setParent;
```
[the remark this answers is not in the document]
Comment (ApolloZhu) [reply lib/AST/ASTScopeCreation.cpp]: I've moved the logic for this case together and make it exit early since the code is well contained and rather short

**Review by DougGregor — APPROVED** [review]

**Pull request merged**